src/Features/Core/Portable/EditAndContinue/DebuggingSession.cs

```diff
@@ -1,20 +1,327 @@
+        /// <returns>
+        /// An MVID and an error message to report, in case an IO exception occurred while reading the binary.
+        /// The MVID is default if either project not built, or an it can't be read from the module binary.
+        /// </returns>
+        public async Task<(Guid Mvid, Diagnostic Error)> GetProjectModuleIdAsync(Project project, CancellationToken cancellationToken)
+        {
+            lock (_projectModuleIdsGuard)
+            {
+                if (_projectModuleIds.TryGetValue(project.Id, out var id))
+                {
+                    return id;
+                }
+            }
+
+            (Guid Mvid, Diagnostic Error) readMvid()
+            {
+                // TODO: use intermediate path
+                var filePath = project.OutputFilePath;
+
+                Contract.ThrowIfFalse(PathUtilities.IsAbsolute(filePath));
+
+                try
+                {
+                    return (ReadModuleVersionId(FileUtilities.OpenRead(filePath)), Error: null);
+                }
+                catch (Exception e) when (e is FileNotFoundException || e is DirectoryNotFoundException)
+                {
+                    return (Mvid: default, Error: null);
+                }
+                catch (Exception e)
+                {
+                    var descriptor = EditAndContinueDiagnosticDescriptors.GetDescriptor(EditAndContinueErrorCode.ErrorReadingFile);
+                    return (Mvid: default, Error: Diagnostic.Create(descriptor, Location.None, new[] { filePath, e.Message }));
+                }
+            }
+
+            var newId = await Task.Run(readMvid, cancellationToken).ConfigureAwait(false);
+
+            lock (_projectModuleIdsGuard)
+            {
+                if (_projectModuleIds.TryGetValue(project.Id, out var id))
+                {
+                    return id;
+                }
+
+                return _projectModuleIds[project.Id] = newId;
+            }
+        }
+
+        internal static Guid ReadModuleVersionId(Stream stream)
+        {
+            using (var reader = new PEReader(stream))
+            {
+                var metadataReader = reader.GetMetadataReader();
+                var mvidHandle = metadataReader.GetModuleDefinition().Mvid;
+                return metadataReader.GetGuid(mvidHandle);
+            }
+        }
+
+        /// <summary>
+        /// Get <see cref="EmitBaseline"/> for given project.
+        /// Must be called on MTA thread.
+        /// </summary>
+        /// <returns>Null if the module corresponding to he project hasn't been loaded yet</returns>
+        /// <exception cref="IOException">Error reading project's binary.</exception>
+        public EmitBaseline TryGetEmitBaseline(ProjectId projectId, Guid mvid)
         {
-            Debug.Assert(initialSolution != null);
-            InitialSolution = initialSolution;
+            Debug.Assert(Thread.CurrentThread.GetApartmentState() == ApartmentState.MTA);
```

Comment:
❗️ This code appears to be making thread affinity demands which are not allowed in layers below EditorFeatures. We need to have a design discussion surrounding this before moving forward.

Reply:
The demands are given by native SymReader implementation, which requires all methods to be called on MTA thread.